### DISCUSSION TITLE
ValidationError customization

### DISCUSSION THREAD

**Discussion by wsaca** [graphql-java/graphql-java#2240] [Q&A]
Hi, are there plans to allow customization of error messages generated as ValidationError?

Currently it is only possible to customize datafetcher exceptions using the DataFetcherExceptionHandler interface. It's also possible to get the error messages of type ValidationError from ExecutionResult, but these messages don't save the information (of the fields or directives) to allow a proper customization.

**Comment by bbakerman** [comment]
We don't have concrete plans to allow "validation message" customisation.

We started but abandoned a plan to at I18N support to the validation messages but the code is not in a great shape in I18N terms ( too much string assembly)

We can of course "improve the messages" that are produced. What is it that you want to change in the validation errors?

Can you give more concrete examples if what we do today versus what you would like to do with them

**Reply by wsaca** [reply]
I understand that modifying the messages to meet the needs of each user is practically impossible, so a design-level solution is a better alternative. I give you the following example:

```json        
{
    "message": "Validation error of type WrongType: argument 'orderBy' with value 'ObjectValue{objectFields=[ObjectField{name='fieldw', value=EnumValue{name='NAME'}}, ObjectField{name='direction', value=EnumValue{name='DESC'}}]}' is missing required fields '[field]' @ 'users'",
    "locations": [
        {
            "line": 2,
            "column": 21
        }
    ],
    "extensions": {
        "classification": "ValidationError"
    }
}
```
We try to normalize all the errors returned to the user, replacing the `message` field with a friendlier one and adding the error information in the` path` and `extensions` fields. I have tried to do it by modifying the errors returned in ExecutionResult but it is not possible to get the necessary information to recreate the error because almost everything is included in the `message` field as text.

**Comment by bbakerman** [comment]
Yeah I agree that is an awful message.  It's a toString() gone mad.

Riffing on the idea that there is a pluggable ValidationErrorBuilding
interface we would need

* An enum that indicate each instance of low level validation error there
can be - there are a LOT
* A way to represent all the input parameters for each error.  This might
need to be a Map<String,Object> kinda thing because there are LOTs of
conditions and hence lots of different parameter conbinations
* A way to plug this interface in - probably on the GraphQL instance itself
* A complete refactor of the very crusty (read old) validation code

The last point is the one I baulk at most.  I have tried in the past to
refactor for I18n (which is a little like the above in many ways) and I
abandoned it based on effort (and my time at the time).  Perhaps its time
to revisit that.

One question I have for you.

If you had a more structured error, with `extensions` and paths and so on.
What would you do with it.  How would a client adjust its behavior because
the error is more structured beyond a message?

**Reply by arnesenfamily** [reply]
Would it be possible / easier to provide a customizable way to handle Exceptions thrown from the DataFetcher.get() method call or a validation failure? The contract of this Exception handler would be required to return a GraphQLError (or a list of them). It seems like this pattern could work similar to how Spring Controller exception handling can be wired in. 

For example, would something like this be easier to include? I know it doesn't address internationalization but it kind of defers that concern to the implementation of the error handler. 
```
package graphql.validation.constraints;

import graphql.PublicSpi;
import graphql.GraphQLError;

/**
 * A GraphQLExceptionHandler
 */
@PublicSpi
public interface GraphQLExceptionHandler<T extends Throwable> {

    /**
     * @return the name of the constraint
     */
    List<GraphQLError> handleException(T error);


}
```

**Reply by bbakerman** [reply]
I think this exists - see `graphql.execution.DataFetcherExceptionHandler`

**Comment by bbakerman** [comment]
Also on a side note - this is quasi related to another feature request
which is "pluggable validation rules".  Some people want to put in custom
rules at pre execution time.

Its also also quasi related to I18n - since if we had pluggable message
formatters, we effectively have I18n capability on those errors

**Comment by ruxiao16** [comment]
is there any plan to add ValidationError customization still? we right now have two types of error response giving back to the client if input are invalid

input modelled by enums -> default behavior 
```json
{
    "errors": [
        {
            "message": "Validation error of type WrongType: argument 'filters.statuses[0]' with value 'EnumValue{name='YOU'}' is not a valid 'PurchaseObjectStatus' - Expected enum literal value not in allowable values -  'EnumValue{name='YOU'}'. @ 'getPurchaseObjects'",
            "locations": [
                {
                    "line": 2,
                    "column": 52
                }
            ],
            "extensions": {
                "classification": "ValidationError"
            }
        }
    ]
}
```

Input that is not modelled by enum but can still fail our customized validation check , we overrides the DataFetcherExceptionHandler so the error response look something like.

 
```json
{
    "errors": [
        {
            "message": "expecting aankoopdossier starts with AD with numbers but input is  AB100129",
            "locations": [],
            "path": [
                "getPurchaseObjects"
            ],
            "extensions": {
                "errorType": "BAD_REQUEST_DATA",
                "errorCode": 4000,
                "debugInfo": {}
            }
        }
    ],
    "data": {
        "getPurchaseObjects": null
    }
}
```

it will be nice to be able to customize ValidationError so the output can be similar to the second case.

**Reply by bbakerman** [reply]
We don't have plans to add a callback during validation - 

However it can be done either in each data fetcher OR via `graphql.execution.instrumentation.Instrumentation#beginExecuteOperation(graphql.execution.instrumentation.parameters.InstrumentationExecuteOperationParameters, graphql.execution.instrumentation.InstrumentationState)`

That is there are callbck spots you can insert custom validation and generate errors and stop execution by throwing `graphql.execution.AbortExecutionException`

**Reply by ruxiao16** [reply]
thanks for the response. Is there a code snippet i can perhaps look at.
